moc_openstack_api_reporting.py

```diff
@@ -0,0 +1,586 @@
+    if event["event_type"] in ceilometer_event_types:
+        ret_list = []
+        delta_time = 0
+        time_0 = datetime.datetime.fromisoformat(event["generated"])
+        for e in ceilometer_event_types[event["event_type"]]:
+            new_event = copy.deepcopy(event)
+            new_event["event_type"] = f"{new_event['event_type']}.{e}"
+            new_event["generated"] = (time_0 + datetime.timedelta(seconds=delta_time)).isoformat()
+            ret_list.append(new_event)
+            delta_time += 10
+        return ret_list
+    return []
+
+
+def compile_server_state(server, project_dict, flavor_dict, user_dict):
+    """This one is implied.  We can tell this from the state of the VM"""
+    launched_ts = None
+    launched_ts = getattr(server, "OS-SRV-USG:launched_at", "")
+    host = None
+    host = getattr(server, "OS-EXT-SRV-ATTR:host", None)
+    terminated_ts = None
+    terminated_ts = getattr(server, "OS-SRV-USG:terminated_at", "")
+    user_name = "unknown user"
+    if server.user_id in user_dict:
+        user_name = user_dict[server.user_id]["name"]
+
+    # I am including all of the commented out fields just to document the entire compute structure
+    # in one spot.
+    server_state = {
+        # "audit_period_beginning": start_ts,   --- only add for compute.instance.exists (either active or deleted)
+        # "audit_period_ending": "end_ts",
+        "disk_gb": str(flavor_dict[server.flavor["id"]]["disk"]),
+        "domain": project_dict[server.tenant_id]["domain_id"],
+        "ephemeral_gb": str(flavor_dict[server.flavor["id"]]["ephemeral"]),
+        # These 2 are filled out below
+        # "event_type": event_type,
+        # "generated": end_ts,
+        "host": host,
+        "instance_id": server.id,
+        "instance_type": flavor_dict[server.flavor["id"]]["name"],
+        # what this should be
+        #   "instance_type_id": server.flavor["id"],
+        # what xdmod is expecting:
+        "instance_type_id": "9",
+        "launched_at": launched_ts,
+        # "deleted_at": terminated_ts,
+        "memory_mb": str(flavor_dict[server.flavor["id"]]["ram"]),
+        # Probably unknowable - I suspect this is what ceilometer adds.
+        # "message_id": "17d6645e-90f9-481e-9751-f8ec3b9397a2",
+        "project_id": server.tenant_id,
+        "project_name": project_dict[server.tenant_id]["name"],
+        "raw": {},
+        # request_id can be NULL
+        # "request_id": "req-00b3079e-8cb1-4e63-aa6f-f96fcbd4771c",
+        # For compute evens, resource_id is the instance id
+        "resource_id": server.id,
+        "root_gb": str(flavor_dict[server.flavor["id"]]["disk"]),
+        "service": "compute",
+        "state": server.status,
+        "tenant_id": server.tenant_id,
+        "user_id": server.user_id,
+        "user_name": user_name,
+        "vcpus": str(flavor_dict[server.flavor["id"]]["vcpus"]),
+    }
+    if terminated_ts is not None:
```

Comment:
At line 390:

`terminated_ts = getattr(server, "OS-SRV-USG:terminated_at", "")`

`terminated_ts` would either bet set to some value or be set to the default `""` which is not `None`. So it looks like we will always hit this branch. Unless of course the attribute could actually return `None`. Am I reading this correctly?

Reply:
yes, naved, I didn't refactor using getattr correctly.